internal/app/upstream/client.go

```diff
@@ -55,11 +73,184 @@ type Response struct {
+	if err != nil {
+		return nil, err
+	}
+
+	ldsClient := v2.NewListenerDiscoveryServiceClient(conn)
+	rdsClient := v2.NewRouteDiscoveryServiceClient(conn)
+	edsClient := v2.NewEndpointDiscoveryServiceClient(conn)
+	cdsClient := v2.NewClusterDiscoveryServiceClient(conn)
+
+	go shutDown(ctx, conn)
+
+	return &client{
+		ldsClient:   ldsClient,
+		rdsClient:   rdsClient,
+		edsClient:   edsClient,
+		cdsClient:   cdsClient,
+		callOptions: callOptions,
+	}, nil
+}
+
+func (m *client) OpenStream(request *v2.DiscoveryRequest) (<-chan *v2.DiscoveryResponse, chan bool, error) {
+	ctx, cancel := context.WithCancel(context.Background())
+	var stream grpc.ClientStream
+	var err error
+	switch request.GetTypeUrl() {
+	case ListenerTypeURL:
+		stream, err = m.ldsClient.StreamListeners(ctx)
+	case ClusterTypeURL:
+		stream, err = m.cdsClient.StreamClusters(ctx)
+	case RouteTypeURL:
+		stream, err = m.rdsClient.StreamRoutes(ctx)
+	case EndpointTypeURL:
+		stream, err = m.edsClient.StreamEndpoints(ctx)
+	default:
+		cancel()
+		return nil, nil, &UnsupportedResourceError{TypeURL: request.GetTypeUrl()}
+	}
+
+	if err != nil {
+		cancel()
+		return nil, nil, err
+	}
+
+	signal := make(chan *version, 1)
+	// The xds protocol https://www.envoyproxy.io/docs/envoy/latest/api-docs/xds_protocol#ack
+	// specifies that the first request be empty nonce and empty version.
+	// The origin server will respond with the latest version.
+	signal <- &version{nonce: "", version: ""}
+
+	done := make(chan bool, 1)
+	response := make(chan *v2.DiscoveryResponse, 1)
+
+	var wg sync.WaitGroup
+	wg.Add(2)
+
+	go send(ctx, cancel, done, &wg, request, response, stream, signal, m.callOptions)
+	go recv(cancel, done, &wg, response, stream, signal)
+
+	go func() {
+		wg.Wait()
+		closeChannels(signal, response)
+	}()
+
+	return response, done, nil
+}
+
+func send(
+	ctx context.Context,
+	cancelFunc context.CancelFunc,
+	done <-chan bool,
+	wg *sync.WaitGroup,
+	request *v2.DiscoveryRequest,
+	response chan *v2.DiscoveryResponse,
+	stream grpc.ClientStream,
+	signal chan *version,
+	callOptions CallOptions) {
+	for {
+		select {
+		case sig, ok := <-signal:
+			if !ok {
+				return
+			}
+			request.ResponseNonce = sig.nonce
+			request.VersionInfo = sig.version
+			err := doWithTimeout(ctx, func() error {
+				return stream.SendMsg(request)
+			}, callOptions.Timeout)
+			if err != nil {
+				select {
+				case <-done:
+					cancelFunc()
+				default:
+					close(response)
+				}
+				wg.Done()
+				return
+			}
+		case <-done:
+			cancelFunc()
+			_ = stream.CloseSend()
+			wg.Done()
+			return
+		}
+	}
+}
+
+func recv(
+	cancelFunc context.CancelFunc,
+	done <-chan bool,
+	wg *sync.WaitGroup,
+	response chan *v2.DiscoveryResponse,
+	stream grpc.ClientStream,
+	signal chan *version) {
+	for {
+		resp := new(v2.DiscoveryResponse)
+		if err := stream.RecvMsg(resp); err != nil {
+			select {
+			case <-done:
+				cancelFunc()
+				wg.Done()
+			default:
+				close(response)
```

Comment:
Similar here, if this is irrecoverable, should we not just issue wg.Done() too? Do we need the select?